### PR TITLE
feat(gateway): add `stopGateway` function

### PR DESCRIPTION
**Pull request by AFink**

### CODE REVIEW

File: gateway/gateway_manager.ts [hunk 1]
```diff
@@ -201,6 +203,8 @@ export interface GatewayManager {
   processGatewayQueue: typeof processGatewayQueue;
   /** Closes shard WebSocket connection properly. */
   closeWS: typeof closeWS;
+  /** Use this function to stop the gateway properly */
```
Comment (itohatweb) [inline gateway/gateway_manager.ts:206]:
```suggestion
  /** Use this function to stop the gateway properly. */
```

File: gateway/stopGateway.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+        clearInterval(shard.heartbeat.intervalId);
+        gateway.closeWS(
+          shard.ws,
+          3061,
+          "Discordeno Testing Finished! Do Not RESUME!",
+        );
+      });
+
+    return await delay(5000);
+}
```
Comment (itohatweb) [inline gateway/stopGateway.ts:14]:
```suggestion
    await delay(5000);
}

```

**Review by itohatweb — CHANGES_REQUESTED** [review]
Looks good overall, just a few nit-picks.

**Review by itohatweb — DISMISSED** [review]

File: gateway/stopGateway.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/** Use this function to stop the gateway properly */
+export async function stopGateway(gateway: GatewayManager) {
```
Comment (Skillz4Killz) [inline gateway/stopGateway.ts:2]:
```suggestion
export async function stopGateway(gateway: GatewayManager, code = 3061, reason = "Discordeno Testing Finished! Do Not RESUME!") {

```

File: gateway/stopGateway.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+/** Use this function to stop the gateway properly */
+export async function stopGateway(gateway: GatewayManager) {
+    // STOP WS
+    gateway.shards.forEach((shard) => {
+        clearInterval(shard.heartbeat.intervalId);
+        gateway.closeWS(
+          shard.ws,
+          3061,
+          "Discordeno Testing Finished! Do Not RESUME!",
```
Comment (Skillz4Killz) [inline gateway/stopGateway.ts:9]:
```suggestion
          reason,
```

File: gateway/stopGateway.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+/** Use this function to stop the gateway properly */
+export async function stopGateway(gateway: GatewayManager) {
+    // STOP WS
+    gateway.shards.forEach((shard) => {
+        clearInterval(shard.heartbeat.intervalId);
+        gateway.closeWS(
+          shard.ws,
+          3061,
```
Comment (Skillz4Killz) [inline gateway/stopGateway.ts:8]:
```suggestion
          code,

```

**Review by Skillz4Killz — APPROVED** [review]

**Pull request merged**